src/sql/workbench/contrib/query/browser/messagesPanelQueryResultsWriter.ts

```diff
@@ -0,0 +1,68 @@
+/*---------------------------------------------------------------------------------------------
+ *  Copyright (c) Microsoft Corporation. All rights reserved.
+ *  Licensed under the Source EULA. See License.txt in the project root for license information.
+ *--------------------------------------------------------------------------------------------*/
+
+import { IResultMessageIntern, Model } from 'sql/workbench/contrib/query/browser/messagePanel';
+import { IQueryMessage, IQueryResultsWriter } from 'sql/workbench/services/query/common/query';
+import QueryRunner from 'sql/workbench/services/query/common/queryRunner';
+import { IDataTreeViewState } from 'vs/base/browser/ui/tree/dataTree';
+import { FuzzyScore } from 'vs/base/common/filters';
+import { Disposable, DisposableStore } from 'vs/base/common/lifecycle';
+import { isArray } from 'vs/base/common/types';
+import { WorkbenchDataTree } from 'vs/platform/list/browser/listService';
+
+export class MessagesPanelQueryResultsWriter extends Disposable implements IQueryResultsWriter {
+	private runner: QueryRunner;
+	private currentUri: string;
+
+	protected queryRunnerDisposables = this._register(new DisposableStore());
+
+	constructor(
+		private readonly model: Model,
+		private readonly tree: WorkbenchDataTree<Model, IResultMessageIntern, FuzzyScore>,
+		private readonly treeStates: Map<string, IDataTreeViewState>,
+	) {
+		super();
+	}
+
+	public subscribeToQueryRunner(): void {
```

Comment:
Instead of having this be in charge of subscribing/unsubscribing to the events - is there any reason why you aren't just making this whole thing disposable and then making each instance tied to a specific query runner? Then you just dispose of it to "unsubscribe" when the editor is closed or a specific output is disabled. 

That'll make the state of this a lot easier to track - instead of relying on stuff like the runner being set/updated correctly